### PR TITLE
Game component methods

### PR DESCRIPTION
**Pull request by riccjohn**
# Main changes to look at

### src/components/GameBoard.tsx

- Changed component to be a stateless functional component since it had no need  for state.

- Most important change here is on line 22, where I pass each `td` an `onClick` method that has a callback containing the correct arguments for the corresponding row and column. This way when a `td` is clicked, it can report back which cell it is, and the game will know where to place a marker.

---

### Src/components/TicTacToe.tsx

- Biggest change is that I instantiate a new game and hold is as a property on the class (line 24) instead of on state.

- I also pass a method to the `<Board />` component so it can deal with a player clicking a cell, and send that data back to this parent component.

#### Questions

I had to define the property on line 6, and typescript wanted it to be set to something, so I initialize the property to a new Game, which then gets overwritten when a user actually starts a game. This seems like a code smell to me, and I'm sure there's a better way, but not sure how.

### CODE REVIEW

File: src/@types/board.d.ts [hunk 1]
```diff
@@ -2,3 +2,13 @@ type PlayerPiece = 'X' | 'O';
 type Square = PlayerPiece | undefined;
 type Row = Square[];
 type Board = Row[];
+
+interface IBoardProps {
```
Comment (ssunday) [inline src/@types/board.d.ts]: I'm of the opinion that prop/state types should be types not interfaces

File: src/components/GameBoard.tsx [hunk 2]
```diff
@@ -1,42 +1,39 @@
...
-  }
-}
+  return (
+    <React.Fragment>
+      {data ? (
+        <div id='board'>
+          <table id='game-table'>
+            <tbody>
+              {data.map((row: Row, rowIdx: number) => (
+                <tr key={rowIdx} className='row'>
```
Comment (rylnd) [inline src/components/GameBoard.tsx]: My preference would just be to spell these out to e.g. `rowIndex`; the potential for confusion or even just having to take 1s to translate it in your head far outweighs the cost of the two characters IMO.

File: src/components/GameBoard.tsx [hunk 3]
```diff
@@ -1,42 +1,39 @@
...
-    );
-  }
-}
+  return (
+    <React.Fragment>
+      {data ? (
+        <div id='board'>
+          <table id='game-table'>
+            <tbody>
+              {data.map((row: Row, rowIdx: number) => (
```
Comment (rylnd) [inline src/components/GameBoard.tsx]: Now that Typescript knows what `data` is, you shouldn't have to type these arguments, they will be inferred correctly.

File: src/components/GameBoard.tsx [hunk 4]
```diff
@@ -1,42 +1,39 @@
...
+                <tr key={rowIdx} className='row'>
+                  {row.map((square: Square, colIdx: number) => (
```
Comment (rylnd) [inline src/components/GameBoard.tsx]: ditto on the type inference.

File: src/components/TicTacToe.tsx [hunk 5]
```diff
@@ -3,11 +3,13 @@ import Game from '../tic-tac-toe';
 import GameBoard from './GameBoard';
 
 class TicTacToe extends React.Component<any, any> {
```
Comment (rylnd) [inline src/components/TicTacToe.tsx]: You should type state and props, here.

File: src/components/TicTacToe.tsx [hunk 6]
```diff
@@ -3,11 +3,13 @@ import Game from '../tic-tac-toe';
...
+  private game: Game = new Game();
+
   constructor(props: {}) {
     super(props);
     this.state = {
-      board: null,
-      boardSize: null,
+      board: undefined,
+      game: undefined,
```
Comment (rylnd) [inline src/components/TicTacToe.tsx]: What happened to `boardSize` here? You're still referencing it below.

File: src/components/GameBoard.tsx [hunk 7]
```diff
@@ -1,42 +1,39 @@
...
+                    <td
+                      key={colIdx}
+                      className='square'
+                      onClick={() => handleClick(rowIdx, colIdx)}
```
Comment (rylnd) [inline src/components/GameBoard.tsx]: Until you're doing something extra here, I would just use the prop:

```jsx
onClick={() => props.handlePlayerInput({ row: rowIdx, col: colIdx })}
```

File: src/components/TicTacToe.tsx [hunk 8]
```diff
@@ -19,9 +21,16 @@ class TicTacToe extends React.Component<any, any> {
 
   public submitBoardSize = (event: any): void => {
```
Comment (ssunday) [inline src/components/TicTacToe.tsx]: This can be typed into a form submit type

Comment (riccjohn) [reply src/components/TicTacToe.tsx]: Do you know where these `React.FormEvent` types are documented? I can only ever find them on StackOverflow.

Comment (ssunday) [reply src/components/TicTacToe.tsx]: I look at the react types file in node modules for it, haha. That's what I use

File: src/components/TicTacToe.tsx [hunk 9]
```diff
@@ -2,26 +2,39 @@ import * as React from 'react';
...
+  game?: Game;
+};
+
+class TicTacToe extends React.Component<object, GameState> {
+  private game: Game = new Game();
+
+  public state = {
+    board: undefined,
+    boardSize: undefined,
+    game: undefined,
```
Comment (rylnd) [inline src/components/TicTacToe.tsx]: I don't think you're using `game` in your state.

**Pull request merged**